### ISSUE TITLE
[feature request]About open in newtab and in background

### ISSUE THREAD

**Issue by NightSharp** [piroor/treestyletab#2523]
I want to open newtab from any actions in newtab and in background.
Such as from urlbar(include searchbar), newtab button of original tabbar and TST, any context menu, bookmark, history and restoring closed tab.

Especially,
About newtab from history, though I set openintab to true in about:config, they still be opened in the current tab unless with middleclick.
About newtab from urlbar/searchbar, newtab button , restored tab and context menu, though I set every item about loadinbackground to true in about:config, they still be opened in foreground.

Actually, I hope that these tab actions about opening in newtab and in background can come true and they can be fully customized and configured in newtab options of TST

----------------------------------------------
Now, though the addon Tab Positon Options can make every newtab be opened in background, It can only work when its option page keep opend so I have to pin this page to the tabbar.
the addon New Tab from Location Bar can make newtab from history be opened in newtab, but it may cause fault to the current tab, such as addon manager page.

**Comment by piroor** [comment]
Hmm, I still negative to add the option to TST, but I'm interested in your motivation and the addon [Tab Position Options](https://addons.mozilla.org/firefox/addon/tab_position_options/).

The design policy of TST is: "compatible to Firefox's native tab bar" + "basic tree management". Because Firefox itself doesn't provide any option to open a new tab in the background from the "+" button, I'm negative about a new option to open new background tab from the "+" button in the sidebar. Instead I hope that something addon who provide the feature works together with TST compatibly.

> It can only work when its option page keep opend so I have to pin this page to the tabbar.

After a research I've realized that it is a bug of the addon itself and can be fixed with this change:

```diff
diff --git a/background.js b/background.js
index a53eedf..348a1cd 100644
--- a/background.js
+++ b/background.js
@@ -241,7 +241,11 @@ function setDefaultValue(){
 	optionobject.loadingurl = [];
 }
 function setOption(optobj){
-	optionobject = optobj;
+	// Ensure that the options object is owned by this background
+	// page namespace, because the caller namespace of this function
+	// (e.g. the options page) may become unloaded and the given
+	// object will become dead.
+	optionobject = JSON.parse(JSON.stringify(optobj));
 	var openpos = optobj.osel;
 	if(openpos){
 		if(openpos == "openbutton1"){
```

I've send a message to the author with the patch, so I hope it become updated in near future, then I can recommend it for people who has motivation same to you.

And I've seen a small incompatibility: the "active tab" state in the sidebar become wrong after the addon works. Such an incompatibility may appear also with other similar addons who control tab focus, so I hope to fix it on TST side if possible.

**Comment by NightSharp** [comment]
Thanks so much



At 2020-03-24 01:29:32, "piroor" <notifications@github.com> wrote:

Hmm, I still negative to add the option to TST, but I'm interested in your motivation and the addon Tab Position Options.

The design policy of TST is: "compatible to Firefox's native tab bar" + "basic tree management". Because Firefox itself doesn't provide any option to open a new tab in the background from the "+" button, I'm negative about a new option to open new background tab from the "+" button in the sidebar. Instead I hope that something addon who provide the feature works together with TST compatibly.

It can only work when its option page keep opend so I have to pin this page to the tabbar.

After a research I've realized that it is a bug of the addon itself and can be fixed with this change:

diff --git a/background.js b/background.js
index a53eedf..348a1cd 100644
--- a/background.js+++ b/background.js@@ -241,7 +241,11 @@ function setDefaultValue(){
 	optionobject.loadingurl = [];
 }
 function setOption(optobj){
-	optionobject = optobj;+	// Ensure that the options object is owned by this background+	// page namespace, because the caller namespace of this function+	// (e.g. the options page) may become unloaded and the given+	// object will become dead.+	optionobject = JSON.parse(JSON.stringify(optobj));
 	var openpos = optobj.osel;
 	if(openpos){
 		if(openpos == "openbutton1"){

I've send a message to the author with the patch, so I hope it become updated in near future, then I can recommend it for people who has motivation same to you.

And I've seen a small incompatibility: the "active tab" state in the sidebar become wrong after the addon works. Such an incompatibility may appear also with other similar addons who control tab focus, so I hope to fix it on TST side if possible.

**Comment by piroor** [comment]
> And I've seen a small incompatibility: the "active tab" state in the sidebar become wrong after the addon works. Such an incompatibility may appear also with other similar addons who control tab focus, so I hope to fix it on TST side if possible.

This has been fixed with the commit 14524d7.

**Comment by irvinm** [comment]
@piroor is there anything more to do with this item?

**Comment by piroor** [comment]
Tab Position Options 5.9 was released at 2020.3.25 with my patch, so the problem "It can only work when its option page keep opend" doesn't happen anymore. Now I can recommend it for people who want to open new tabs in the background.
